File: ibm/src/go/filesystem/filesystem.go
package main

import (
    "strconv"
    "io/ioutil"
    "log"
    "time"
    "os"
    "math/rand"
)

func Main(params map[string]interface{}) map[string]interface{} {

    rnd := rand.Intn(900000) + 100000

    if _, err := os.Stat("/tmp/test"); os.IsNotExist(err) {
        os.Mkdir("/tmp/test", 0777)
    }

    if _, err := os.Stat("/tmp/test/"+strconv.Itoa(rnd)); os.IsNotExist(err) {
        os.Mkdir("/tmp/test/"+strconv.Itoa(rnd), 0777)
    }

    buf1, err := ioutil.ReadFile("/proc/self/cgroup")
	if err != nil {
		log.Fatal(err)
    }
    instance_id := string(buf1)

    buf2, err := ioutil.ReadFile("/proc/cpuinfo")
	if err != nil {
		log.Fatal(err)
    }
    cpuinfo := string(buf2)

    buf3, err := ioutil.ReadFile("/proc/meminfo")
	if err != nil {
		log.Fatal(err)
    }
    meminfo := string(buf3)

    buf4, err := ioutil.ReadFile("/proc/uptime")
	if err != nil {
		log.Fatal(err)
    }
    uptime := string(buf4)

    buf5, err := ioutil.ReadFile("/sys/class/dmi/id/product_uuid")
	if err != nil {
		log.Fatal(err)
    }
    machine_id := string(buf5)

    n := 10000
    size := 10240

    if _, ok := params["n"]; ok {
        n, err = strconv.Atoi(params["n"].(string))
        if err != nil {
            log.Fatal(err)
        }
    } else {
        n = 10000
    }

    if _, ok := params["size"]; ok {
        size, err = strconv.Atoi(params["size"].(string))
        if err != nil {
            log.Fatal(err)
        }
    } else {
        size = 10240
    }

    var text string = ""
    
    for i := 0; i < size; i++ {
        text += "A"
    }

    startWrite := time.Now()
    for i := 0; i < n; i++ {
        f, err := os.Create("/tmp/test/" + strconv.Itoa(rnd) + "/" + strconv.Itoa(i) + ".txt")
        if err != nil {
            log.Fatal(err)
        }
        f.WriteString(text)
        f.Close()
    }
    elapsedWrite := time.Since(startWrite)
    
    var test string = ""

    startRead := time.Now()
    for i := 0; i < n; i++ {
        buf, err := ioutil.ReadFile("/tmp/test/" + strconv.Itoa(rnd) + "/" + strconv.Itoa(i) + ".txt")
	    if err != nil {
		    log.Fatal(err)
        }
        test = string(buf)
    }
    elapsedRead := time.Since(startRead)

    log.Print(test, elapsedWrite, elapsedRead)

    files, err := ioutil.ReadDir("/tmp/test/" + strconv.Itoa(rnd))
    if err != nil {
        log.Fatal(err)
    }

    if _, err := os.Stat("/tmp/test/"+strconv.Itoa(rnd)); !os.IsNotExist(err) {
        os.RemoveAll("/tmp/test/"+strconv.Itoa(rnd))
    }

    msg := map[string]interface{}{}
    msg["success"] = len(files) == n
    msg["payload"] = map[string]interface{}{}
    msg["payload"].(map[string]interface{})["test"] = "filesystem test"
    msg["payload"].(map[string]interface{})["n"] = len(files)
    msg["payload"].(map[string]interface{})["size"] = size
    msg["payload"].(map[string]interface{})["timewrite"] = strconv.FormatInt(int64(elapsedWrite / time.Millisecond), 10)
    msg["payload"].(map[string]interface{})["timeread"] = strconv.FormatInt(int64(elapsedRead / time.Millisecond), 10)
    msg["payload"].(map[string]interface{})["time"] = strconv.FormatInt(int64(elapsedWrite / time.Millisecond) + int64(elapsedRead / time.Millisecond), 10)
    msg["metrics"] = map[string]string{}
    msg["metrics"].(map[string]string)["machineid"] = machine_id
    msg["metrics"].(map[string]string)["instanceid"] = instance_id
    msg["metrics"].(map[string]string)["cpu"] = cpuinfo
    msg["metrics"].(map[string]string)["mem"] = meminfo
    msg["metrics"].(map[string]string)["uptime"] = uptime

    return msg

 }

File: ibm/src/dotnet/Filesystem/Filesystem.cs
using System;
using System.IO;
using System.Diagnostics;
using Newtonsoft.Json.Linq;

namespace Filesystem
{
    public class FilesystemDotnet
    {
        public JObject Main(JObject args)
        {
            Random random = new Random();
            int rnd = random.Next(100000, 1000000);

            if(!Directory.Exists("/tmp/test")) {
                System.IO.Directory.CreateDirectory("/tmp/test");
            }

            if(!Directory.Exists("/tmp/test/" + rnd.ToString())) {
                System.IO.Directory.CreateDirectory("/tmp/test/" + rnd.ToString());
            }
            
            string machineId = File.ReadAllText("/sys/class/dmi/id/product_uuid");
            string instanceId = File.ReadAllText("/proc/self/cgroup");
            string cpuinfo = File.ReadAllText("/proc/cpuinfo");
            string meminfo = File.ReadAllText("/proc/meminfo");
            string uptime = File.ReadAllText("/proc/uptime");

            int n = 10000;
            int size = 10240;

            if(args["n"] != null) {
                bool parseOk = Int32.TryParse(args["n"].ToString(), out n);
                if(!parseOk) {
                    n = 10000;
                }
            } else {
                n = 10000;
            }

            if(args["size"] != null) {
                bool parseOk = Int32.TryParse(args["size"].ToString(), out size);
                if(!parseOk) {
                    size = 10240;
                }
            } else {
                size = 10240;
            }

            string text = "";

            for(short i = 0; i<size; i++) {
                text += "A";
            }

            Stopwatch swWrite = new Stopwatch();
            swWrite.Start();
            for(short i = 0; i<n; i++) {
                File.WriteAllText("/tmp/test/"+rnd.ToString()+"/"+i+".txt", text);
            }
            swWrite.Stop();

            Stopwatch swRead = new Stopwatch();
            swRead.Start();
            for(short i = 0; i<n; i++) {
                string test = File.ReadAllText("/tmp/test/"+rnd.ToString()+"/"+i+".txt");
            }
            swRead.Stop();

            string[] files = Directory.GetFiles("/tmp/test/"+rnd.ToString());

            if(Directory.Exists("/tmp/test/"+rnd.ToString())) {
                System.IO.Directory.Delete("/tmp/test/"+rnd.ToString(), true);
            }

            JObject message = new JObject();
            message.Add("success", new JValue((files.Length == n)));
            JObject payload = new JObject();
            payload.Add("test", new JValue("filesystem test"));
            payload.Add("n", new JValue(files.Length));
            payload.Add("size", new JValue(size));
            payload.Add("timewrite", new JValue(swWrite.Elapsed.TotalMilliseconds));
            payload.Add("timeread", new JValue(swRead.Elapsed.TotalMilliseconds));
            payload.Add("time", new JValue(swWrite.Elapsed.TotalMilliseconds+swRead.Elapsed.TotalMilliseconds));
            message.Add("payload", payload);
            JObject metrics = new JObject();
            metrics.Add("machineid", new JValue(string.Join("\n", machineId)));
            metrics.Add("instanceid", new JValue(string.Join("\n", instanceId)));
            metrics.Add("cpu", new JValue(string.Join("\n", cpuinfo)));
            metrics.Add("mem", new JValue(string.Join("\n", meminfo)));
            metrics.Add("uptime", new JValue(string.Join("\n", uptime)));
            message.Add("metrics", metrics);

            return (message);
            
        }
    }
 }